lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -447,11 +447,14 @@ static bool HasReflectionInfo(ObjectFile *obj_file) {
   return hasReflectionSection;
 }
 
-SwiftLanguageRuntimeImpl::ReflectionContextInterface *
+SwiftLanguageRuntimeImpl::ThreadSafeReflectionContext 
 SwiftLanguageRuntimeImpl::GetReflectionContext() {
+  {
+    std::lock_guard<std::recursive_mutex> lock(m_add_module_mutex);
   if (!m_initialized_reflection_ctx)
-    SetupReflection();
-  return m_reflection_ctx.get();
+        SetupReflection();
+  }
+  return {m_reflection_ctx.get(), m_add_module_mutex};
```

Comment:
The mutex is locked, unlocked, then locked again. The first lock is at the start of the `{}` block scope, and it's then unlocked at the end of the block scope. The second lock happens in the `ThreadSafeReflectionContext` constructed for the return statement.

The result is the lock is not owned/locked continuously. Should it be? Could there be a race, and unexpected behavior, if some other caller locks the mutex in the window where it's unlocked?

Reply:
I believe this is be safe. The mutex is locked until ReflectionContext is initialized in thread 1, after that, it could be that thread 2 acquires the mutex before thread 1 re-acquires it. But at that point ReflectionContext is already initialized correctly, and whatever operation that needs the ReflectionContext in thread 1 can wait until thread 2 finishes its work.

 By the way, this is necessary because locks aren't movable or copiable.

Reply:
> By the way, this is necessary because locks aren't movable or copiable.

I think you could work around this by creating the object first and relying on it holding the relevant lock?

Reply:
@JDevlieghere I thought that too, but it doesn't work. Probably because RVO is an optimization and doesn't affect the semantics (even though it's guaranteed in this case).